// @pythnetwork/pyth-lazer-sdk/dist/cjs/protocol.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BINARY_UPDATE_FORMAT_MAGIC_LE () {
        return BINARY_UPDATE_FORMAT_MAGIC_LE;
    },
    get CustomSocketClosureCodes () {
        return CustomSocketClosureCodes;
    },
    get FORMAT_MAGICS_LE () {
        return FORMAT_MAGICS_LE;
    }
});
const BINARY_UPDATE_FORMAT_MAGIC_LE = 461_928_307;
const FORMAT_MAGICS_LE = {
    JSON: 3_302_625_434,
    EVM: 2_593_727_018,
    SOLANA: 2_182_742_457,
    LE_ECDSA: 1_296_547_300,
    LE_UNSIGNED: 1_499_680_012
};
var CustomSocketClosureCodes = /*#__PURE__*/ function(CustomSocketClosureCodes) {
    CustomSocketClosureCodes[CustomSocketClosureCodes["CLIENT_TIMEOUT_BUT_RECONNECTING"] = 4000] = "CLIENT_TIMEOUT_BUT_RECONNECTING";
    return CustomSocketClosureCodes;
}({});
