// @pythnetwork/pyth-lazer-sdk/dist/cjs/socket/resilient-websocket.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ResilientWebSocket", {
    enumerable: true,
    get: function() {
        return ResilientWebSocket;
    }
});
const _isomorphicws = /*#__PURE__*/ _interop_require_default(require("isomorphic-ws"));
const _tslog = require("ts-log");
const _protocol = require("../protocol.cjs");
const _envutil = require("../util/env-util.cjs");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const DEFAULT_HEARTBEAT_TIMEOUT_DURATION_MS = 5000; // 5 seconds
const DEFAULT_MAX_RETRY_DELAY_MS = 1000; // 1 second'
const DEFAULT_LOG_AFTER_RETRY_COUNT = 10;
class ResilientWebSocket {
    endpoint;
    wsOptions;
    logger;
    heartbeatTimeoutDurationMs;
    maxRetryDelayMs;
    logAfterRetryCount;
    wsClient;
    wsUserClosed = false;
    wsFailedAttempts;
    heartbeatTimeout;
    retryTimeout;
    _isReconnecting = false;
    isReconnecting() {
        return this._isReconnecting;
    }
    isConnected() {
        return this.wsClient?.readyState === _isomorphicws.default.OPEN;
    }
    shouldLogRetry() {
        return this.wsFailedAttempts % this.logAfterRetryCount === 0;
    }
    onError;
    onMessage;
    onReconnect;
    constructor(config){
        this.endpoint = config.endpoint;
        this.wsOptions = config.wsOptions;
        this.logger = config.logger ?? _tslog.dummyLogger;
        this.heartbeatTimeoutDurationMs = config.heartbeatTimeoutDurationMs ?? DEFAULT_HEARTBEAT_TIMEOUT_DURATION_MS;
        this.maxRetryDelayMs = config.maxRetryDelayMs ?? DEFAULT_MAX_RETRY_DELAY_MS;
        this.logAfterRetryCount = config.logAfterRetryCount ?? DEFAULT_LOG_AFTER_RETRY_COUNT;
        this.wsFailedAttempts = 0;
        this.onError = (error)=>{
            void error;
        };
        this.onMessage = (data)=>{
            void data;
        };
        this.onReconnect = ()=>{
        // Empty function, can be set by the user.
        };
    }
    send(data) {
        this.logger.debug(`Sending message`);
        if (this.isConnected()) {
            this.wsClient.send(data);
        } else {
            this.logger.warn(`WebSocket to ${this.endpoint} is not connected. Cannot send message.`);
        }
    }
    startWebSocket() {
        if (this.wsUserClosed) {
            this.logger.error("Connection was explicitly closed by user. Will not reconnect.");
            return;
        }
        if (this.wsClient !== undefined) {
            this.logger.info("WebSocket client already started.");
            return;
        }
        if (this.wsFailedAttempts == 0) {
            this.logger.info(`Creating Web Socket client`);
        }
        if (this.retryTimeout !== undefined) {
            clearTimeout(this.retryTimeout);
            this.retryTimeout = undefined;
        }
        // browser constructor supports a different 2nd argument for the constructor,
        // so we need to ensure it's not included if we're running in that environment:
        // https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#protocols
        this.wsClient = new _isomorphicws.default(this.endpoint, (0, _envutil.envIsBrowserOrWorker)() ? undefined : this.wsOptions);
        this.wsClient.addEventListener("open", ()=>{
            this.logger.info("WebSocket connection established");
            this.wsFailedAttempts = 0;
            this._isReconnecting = false;
            this.resetHeartbeat();
            this.onReconnect();
        });
        this.wsClient.addEventListener("close", (e)=>{
            if (this.wsUserClosed) {
                this.logger.info(`WebSocket connection to ${this.endpoint} closed by user`);
            } else {
                if (this.shouldLogRetry()) {
                    this.logger.warn(`WebSocket connection to ${this.endpoint} closed unexpectedly: Code: ${e.code.toString()}`);
                }
                this.handleReconnect();
            }
        });
        this.wsClient.addEventListener("error", (event)=>{
            this.onError(event);
        });
        this.wsClient.addEventListener("message", (event)=>{
            this.resetHeartbeat();
            this.onMessage(event.data);
        });
        if ("on" in this.wsClient) {
            this.wsClient.on("ping", ()=>{
                this.logger.info("Ping received");
                this.resetHeartbeat();
            });
        }
    }
    resetHeartbeat() {
        if (this.heartbeatTimeout !== undefined) {
            clearTimeout(this.heartbeatTimeout);
        }
        this.heartbeatTimeout = setTimeout(()=>{
            const warnMsg = "Connection timed out. Reconnecting...";
            this.logger.warn(warnMsg);
            if (this.wsClient) {
                if (typeof this.wsClient.terminate === "function") {
                    this.wsClient.terminate();
                } else {
                    // terminate is an implementation detail of the node-friendly
                    // https://www.npmjs.com/package/ws package, but is not a native WebSocket API,
                    // so we have to use the close method
                    this.wsClient.close(_protocol.CustomSocketClosureCodes.CLIENT_TIMEOUT_BUT_RECONNECTING, warnMsg);
                }
            }
            this.handleReconnect();
        }, this.heartbeatTimeoutDurationMs);
    }
    handleReconnect() {
        if (this.wsUserClosed) {
            this.logger.info("WebSocket connection closed by user, not reconnecting.");
            return;
        }
        if (this.heartbeatTimeout !== undefined) {
            clearTimeout(this.heartbeatTimeout);
        }
        if (this.retryTimeout !== undefined) {
            clearTimeout(this.retryTimeout);
        }
        this.wsFailedAttempts += 1;
        this.wsClient = undefined;
        this._isReconnecting = true;
        if (this.shouldLogRetry()) {
            this.logger.error("Connection closed unexpectedly or because of timeout. Reconnecting after " + String(this.retryDelayMs()) + "ms.");
        }
        this.retryTimeout = setTimeout(()=>{
            this.startWebSocket();
        }, this.retryDelayMs());
    }
    closeWebSocket() {
        if (this.wsClient !== undefined) {
            this.wsClient.close();
            this.wsClient = undefined;
        }
        this.wsUserClosed = true;
    }
    /**
   * Calculates the delay in milliseconds for exponential backoff based on the number of failed attempts.
   *
   * The delay increases exponentially with each attempt, starting at 20ms for the first attempt,
   * and is capped at maxRetryDelayMs for attempts greater than or equal to 10.
   *
   * @returns The calculated delay in milliseconds before the next retry.
   */ retryDelayMs() {
        if (this.wsFailedAttempts >= 10) {
            return this.maxRetryDelayMs;
        }
        return Math.min(2 ** this.wsFailedAttempts * 10, this.maxRetryDelayMs);
    }
}
