// @pythnetwork/pyth-lazer-sdk/dist/cjs/socket/websocket-pool.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "WebSocketPool", {
    enumerable: true,
    get: function() {
        return WebSocketPool;
    }
});
const _ttlcache = /*#__PURE__*/ _interop_require_default(require("@isaacs/ttlcache"));
const _tslog = require("ts-log");
const _resilientwebsocket = require("./resilient-websocket.cjs");
const _constants = require("../constants.cjs");
const _index = require("../emitter/index.cjs");
const _index1 = require("../util/index.cjs");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const DEFAULT_NUM_CONNECTIONS = 4;
class WebSocketPool extends _index.IsomorphicEventEmitter {
    logger;
    rwsPool;
    cache;
    subscriptions;
    messageListeners;
    allConnectionsDownListeners;
    wasAllDown = true;
    checkConnectionStatesInterval;
    constructor(logger){
        super(), this.logger = logger;
        this.rwsPool = [];
        this.cache = new _ttlcache.default({
            ttl: 1000 * 10
        }); // TTL of 10 seconds
        this.subscriptions = new Map();
        this.messageListeners = [];
        this.allConnectionsDownListeners = [];
        // Start monitoring connection states
        this.checkConnectionStatesInterval = setInterval(()=>{
            this.checkConnectionStates();
        }, 100);
    }
    /**
   * Creates a new WebSocketPool instance that uses multiple redundant WebSocket connections for reliability.
   * Usage semantics are similar to using a regular WebSocket client.
   * @param urls - List of WebSocket URLs to connect to
   * @param token - Authentication token to use for the connections
   * @param numConnections - Number of parallel WebSocket connections to maintain (default: 3)
   * @param logger - Optional logger to get socket level logs. Compatible with most loggers such as the built-in console and `bunyan`.
   */ static async create(config, token, logger) {
        const urls = config.urls ?? [
            _constants.DEFAULT_STREAM_SERVICE_0_URL,
            _constants.DEFAULT_STREAM_SERVICE_1_URL
        ];
        const log = logger ?? _tslog.dummyLogger;
        const pool = new WebSocketPool(log);
        const numConnections = config.numConnections ?? DEFAULT_NUM_CONNECTIONS;
        // bind a handler to capture any emitted errors and send them to the user-provided
        // onWebSocketPoolError callback (if it is present)
        if (typeof config.onWebSocketPoolError === "function") {
            pool.on("error", config.onWebSocketPoolError);
            pool.once("shutdown", ()=>{
                // unbind all error handlers so we don't leak memory
                pool.off("error");
            });
        }
        for(let i = 0; i < numConnections; i++){
            const baseUrl = urls[i % urls.length];
            const isBrowser = (0, _index1.envIsBrowserOrWorker)();
            const url = isBrowser ? (0, _index1.addAuthTokenToWebSocketUrl)(baseUrl, token) : baseUrl;
            if (!url) {
                throw new Error(`URLs must not be null or empty`);
            }
            const wsOptions = {
                ...config.rwsConfig?.wsOptions,
                headers: isBrowser ? undefined : {
                    Authorization: `Bearer ${token}`
                }
            };
            const rws = new _resilientwebsocket.ResilientWebSocket({
                ...config.rwsConfig,
                endpoint: url,
                wsOptions,
                logger: log
            });
            // If a websocket client unexpectedly disconnects, ResilientWebSocket will reestablish
            // the connection and call the onReconnect callback.
            rws.onReconnect = ()=>{
                if (rws.wsUserClosed) {
                    return;
                }
                for (const [, request] of pool.subscriptions){
                    try {
                        rws.send(JSON.stringify(request));
                    } catch (error) {
                        pool.logger.error("Failed to resend subscription on reconnect:", error);
                    }
                }
            };
            // eslint-disable-next-line @typescript-eslint/no-deprecated
            const onErrorHandler = config.onWebSocketError ?? config.onError;
            if (typeof onErrorHandler === "function") {
                rws.onError = onErrorHandler;
            }
            // Handle all client messages ourselves. Dedupe before sending to registered message handlers.
            rws.onMessage = (data)=>{
                pool.dedupeHandler(data).catch((error)=>{
                    pool.emitPoolError(error, "Error in WebSocketPool dedupeHandler");
                });
            };
            pool.rwsPool.push(rws);
            rws.startWebSocket();
        }
        pool.logger.info(`Started WebSocketPool with ${numConnections.toString()} connections. Waiting for at least one to connect...`);
        while(!pool.isAnyConnectionEstablished()){
            await new Promise((resolve)=>setTimeout(resolve, 100));
        }
        pool.logger.info(`At least one WebSocket connection is established. WebSocketPool is ready.`);
        return pool;
    }
    emitPoolError(error, context) {
        const err = error instanceof Error ? error : new Error(String(error));
        if (context) {
            this.logger.error(context, err);
        } else {
            this.logger.error("WebSocketPool error", err);
        }
        for (const errHandler of this.getListeners("error")){
            try {
                errHandler(err);
            } catch (handlerError) {
                this.logger.error("WebSocketPool error handler threw", handlerError);
            }
        }
    }
    /**
   * Checks for error responses in JSON messages and throws appropriate errors
   */ handleErrorMessages(data) {
        const message = JSON.parse(data);
        if (message.type === "subscriptionError") {
            throw new Error(`Error occurred for subscription ID ${String(message.subscriptionId)}: ${message.error}`);
        } else if (message.type === "error") {
            throw new Error(`Error: ${message.error}`);
        }
    }
    async constructCacheKeyFromWebsocketData(data) {
        if (typeof data === "string") return data;
        const buff = await (0, _index1.bufferFromWebsocketData)(data);
        return buff.toString("hex");
    }
    /**
   * Handles incoming websocket messages by deduplicating identical messages received across
   * multiple connections before forwarding to registered handlers
   */ dedupeHandler = async (data)=>{
        const cacheKey = await this.constructCacheKeyFromWebsocketData(data);
        if (this.cache.has(cacheKey)) {
            this.logger.debug("Dropping duplicate message");
            return;
        }
        this.cache.set(cacheKey, true);
        if (typeof data === "string") {
            this.handleErrorMessages(data);
        }
        try {
            await Promise.all(this.messageListeners.map((handler)=>handler(data)));
        } catch (error) {
            this.emitPoolError(error, "Error in WebSocketPool message handler");
        }
    };
    sendRequest(request) {
        for (const rws of this.rwsPool){
            try {
                rws.send(JSON.stringify(request));
            } catch (error) {
                this.emitPoolError(error, "Failed to send WebSocket request");
            }
        }
    }
    addSubscription(request) {
        if (request.type !== "subscribe") {
            this.emitPoolError(new Error("Request must be a subscribe request"));
            return;
        }
        this.subscriptions.set(request.subscriptionId, request);
        this.sendRequest(request);
    }
    removeSubscription(subscriptionId) {
        this.subscriptions.delete(subscriptionId);
        const request = {
            type: "unsubscribe",
            subscriptionId
        };
        this.sendRequest(request);
    }
    addMessageListener(handler) {
        this.messageListeners.push(handler);
    }
    /**
   * Calls the handler if all websocket connections are currently down or in reconnecting state.
   * The connections may still try to reconnect in the background.
   */ addAllConnectionsDownListener(handler) {
        this.allConnectionsDownListeners.push(handler);
    }
    areAllConnectionsDown() {
        return this.rwsPool.every((ws)=>!ws.isConnected() || ws.isReconnecting());
    }
    isAnyConnectionEstablished() {
        return this.rwsPool.some((ws)=>ws.isConnected());
    }
    checkConnectionStates() {
        const allDown = this.areAllConnectionsDown();
        // If all connections just went down
        if (allDown && !this.wasAllDown) {
            this.wasAllDown = true;
            this.logger.error("All WebSocket connections are down or reconnecting");
            // Notify all listeners
            for (const listener of this.allConnectionsDownListeners){
                try {
                    listener();
                } catch (error) {
                    this.emitPoolError(error, "All-connections-down listener threw");
                }
            }
        }
        // If at least one connection was restored
        if (!allDown && this.wasAllDown) {
            this.wasAllDown = false;
        }
    }
    shutdown() {
        for (const rws of this.rwsPool){
            rws.closeWebSocket();
        }
        this.rwsPool = [];
        this.subscriptions.clear();
        this.messageListeners = [];
        this.allConnectionsDownListeners = [];
        clearInterval(this.checkConnectionStatesInterval);
        // execute all bound shutdown handlers
        for (const shutdownHandler of this.getListeners("shutdown")){
            try {
                shutdownHandler();
            } catch (error) {
                this.emitPoolError(error, "Shutdown handler threw");
            }
        }
        // ensure any error handlers are removed to avoid leaks
        this.off("error");
    }
}
